### PR TITLE
docs(cli): Improve CLI documentation structure

### PR DESCRIPTION
**Pull request by santiagogaray**
Hi @chriswmackey,

I believe I got this to a point decent enough for a first pass. Without having discussed any of the details on what we are after. following closely the queenbee approach as you suggested really makes sense as a starting point. 
Also, I did a few minor changes to the main documentation page, so take a look at that too if you can.

Hope this is good to get going, feel free to take a look whenever you have some time!

### CODE REVIEW

**Comment by chriswmackey** [comment]
Thanks @santiagogaray . That was fast!

I'l try to review this tomorrow once I get the chance.

File: docs/result.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Result
```
Comment (chriswmackey) [inline docs/result.rst:1]: If we need to have all of these .rst files in the repo, could we put them in their own `cli` folder like on the queenbee repo?
https://github.com/ladybug-tools/queenbee/tree/master/docs/cli

File: docs/lib.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Library
```
Comment (chriswmackey) [inline docs/lib.rst:1]: Is there any way to have these .rst files generated by Sphinx somehow?  I feel that this is essentially what is happening when we call the following command:
`sphinx-apidoc -f -e -d 4 -o ./docs ./honeybee_energy`
I guess adding a new .rst file every time that we add a new CLI group isn't that bad but it would be good to make sure there are no automated options for this since we will definitely be adding more groups in the future and (probably) forgetting to add an rst file here.

File: docs/api.rst [hunk 3]
```diff
@@ -0,0 +1,14 @@
+API Docs
```
Comment (chriswmackey) [inline docs/api.rst:1]: Why is this file called `api.rst` and not `modules.rst`?

File: docs/lib.rst (hunk 2 above)
Comment (chriswmackey) [reply docs/lib.rst:1]: On a related note, does this mean that the command we have to run to generate the docs is now:
```
sphinx-apidoc -f -e -d 4 -o ./docs/api ./honeybee_energy
```
If so, can you update the ReadMe.MD to reflect this?

File: docs/result.rst (hunk 1 above)
Comment (santiagogaray) [reply docs/result.rst:1]: Yes, good point that would be cleaner!

File: docs/lib.rst (hunk 2 above)
Comment (santiagogaray) [reply docs/lib.rst:1]: I agree, you nailed it! It would be great to auto generating the all the cli rst files for each group. So the bad news is that it is not as straightforward as the sphinx -e option that separates modules, nothing is clearly available as an option within the sphinx-cli extension at least, 
That said, I'm not ready to give up, I will like to look into it deeper if that's okay. If a simple approach is not possible we can look at less straightforward ways to get there. I guess the last resource would be to try to generate these files with a routine inside the conf.py (can only try!).
 
On your second question, the command should stay as usual (without the /api), Also we can eliminate the api.rst file if better (see below).

File: docs/api.rst (hunk 3 above)
Comment (santiagogaray) [reply docs/api.rst:1]: So this was done for the sole purpose of having the API Docs title in the sidebar and in the index page at the bottom (the api.rst connects to the module.rst). I agree this may not merit an additional rst file. 
I was following the queenbee approach though I think that library is very unique in itself, so we can keep it simpler here and avoid the api.rst file if you prefer.
Below is how it will look like, Basically I generated the doc index page with the module.rst instead of the api.rst.

image

**Comment by santiagogaray** [comment]
Thanks @chriswmackey for taking time to look into the PR changes and adding some questions. I made some comments on them for your review.

File: docs/lib.rst (hunk 2 above)
Comment (chriswmackey) [reply docs/lib.rst:1]: Sounds good and let me know if you find anything. Manually adding the `.rst` files is definitely not the end of the world and it's an acceptable solution if it's the best that we find. Still, if we are going to deploy a set of practices across several repos, it seems worthwhile to invest some time up-front in figuring out what the best setup is.

File: docs/api.rst (hunk 3 above)
Comment (chriswmackey) [reply docs/api.rst:1]: I see now. That landing page looks beautiful and using `api.rst` is a good idea.

File: docs/lib.rst (hunk 2 above)
Comment (santiagogaray) [reply docs/lib.rst:1]: I agree, will focus on this and will keep you posted!

**Comment by santiagogaray** [comment]
Hi @chriswmackey 
First of all, I apologise for taking this long to get back to you with progress. We had a week-long invitation to the cape that my family could not resist. I'm a lot more available from now on so I'm pretty sure we'll get everything before the the of the month.

So I did have enough time to look deeper at the automated .rst file creation, looking at different ways to achieve this:

1- First I looked at the sphinx-click plug-in, and unfortunately does not offer any api-level access,, which I believe leaves us with  the only option to modify the code to create/use our own version of the plug-in. Did not want to go this path until we discussed this further.

2- Also tried using the sphinx api to somehow direct sphinx to detect the rst files  Even though some of sphinx's core events give us access to the file environment, unfortunately sphinx does not allow users to add new files to it at that point in the process, Not sure if we can create a new parser or something, but that may take some time to figure out, probably better to modify the sphinx-click plugin.

3- The simpler approach was doing my own reading/parsing the cli folder by creating a routine inside the conf.py with the steps below.  
  - read the \cli directory and extract the existing modules
  - write the .rst files for each module
  - update the cli\index file to include the new commands (assumes each module is a command group)
This did work for this library but if we go this route I will like to test it out and extend it to make sure that it covers all edge cases and that it works for all current and future libraries.

I hope this is clear. I'm sending the updated code below. Please let me know your thought when you get a chance. I'm definitely open to any suggestions. Sorry again for the late update!

**Comment by chriswmackey** [comment]
Thanks @santiagogaray  and no worries. Give me a chance to read what you wrote with fresh eyes and I'll post some feedback tomorrow or Sunday.

File: docs/conf.py [hunk 4]
```diff
@@ -225,6 +225,175 @@
...
+# differ from their library name. Format: {library_name : command_name}
+ht_lib_command = {}
+
+# Repository-library hash table. Creared to address library names that
+#  differ from their repository name (beyond the dash-lower dash difference).
+# Format: {repo_name : library_name}
+ht_repo_lib = {}
+
+
+def create_CLI_files():
```
Comment (chriswmackey) [inline docs/conf.py:248]: `cli` should be lower case to conform with PEP8.

**Comment by chriswmackey** [comment]
FYI, @santiagogaray , this is what the whole core library dependency tree looks like right now:
Ladybug Tools Repo Structures (7)
If you see any other repos in this tree that we missed the first time around, I think we'd like to add them to the new scope if possible.  Depending on how large or a scope you want to take on, we also have 5 Grasshopper repos that I know we eventually want to document but this would be with a completely different setup from the sphinx workflows we use in the core libraries:
Ladybug Tools Repo Structures (6)
Specifically, we would like to make some [gitbook documentation like what you see we did here for our ladybug legacy plugins](https://mostapharoudsari.gitbooks.io/ladybug-primer/content/). Mostapha already started [a repo for the newest ladybug-grasshopper components](https://github.com/ladybug-tools/ladybug-primer) and [corresponding gitbook that is generated from the .MD files on that repo](https://docs.ladybug.tools/ladybug-primer/) but we have not updated it in a while. We have a (currently hacky) Grasshopper script that we use to drop all of the components on a canvass and take a screenshot of them, which I can share with you if you are interested to take this type of documentation on. Just let me know.

File: docs/conf.py [hunk 5]
```diff
@@ -225,6 +225,175 @@
...
+    """Generates the rst files with CLI directives for each module detected in
+    the library's cli folder. Updates index.rst file to include a list of
+    sub-command documenation pages.
+    """
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py:252]: Good point, done!

File: docs/conf.py [hunk 6]
```diff
@@ -225,6 +225,175 @@
...
+    if not result:
+        print ("Something went wrong during CLI index update")
+
+    return
+
+
+def get_CLI_data(repo_path):
+    """Given a repository path, retrieves the repository and library names along
+    with the names of modules found inside the cli folder.
+    """
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py:292]: Great point, done!

File: docs/conf.py [hunk 7]
```diff
@@ -225,6 +225,175 @@
...
+    lib_path = os.path.abspath(os.path.join(repo_path, lib_name))
+
+    if not os.path.isdir(lib_path):
+        print ("Cannot find library path")
+        return None
+
+    CLI_path = os.path.join(lib_path, 'cli')
+    if not os.path.isdir(lib_path):
+        print ("No CLI library found")
+        return None
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py:304]: Glad to hear!

File: docs/conf.py [hunk 8]
```diff
@@ -225,6 +225,175 @@
...
+    """
+    repo_name = os.path.split(repo_path)[1]
+    lib_name = repo_name.replace("-", "_")
+    lib_path = os.path.abspath(os.path.join(repo_path, lib_name))
+
+    if not os.path.isdir(lib_path):
+        print ("Cannot find library path")
+        return None
+
+    CLI_path = os.path.join(lib_path, 'cli')
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py:301]: Yes, thanks for reminding me this. changed also in function names.

File: docs/conf.py (hunk 4 above)
Comment (santiagogaray) [reply docs/conf.py:248]: Done!

File: docs/conf.py [hunk 9]
```diff
@@ -225,6 +225,175 @@
...
+        doc_folder: The path where cli docs will be saved.
+    """
+
+    # Creating missing CLI rst files
+    print ("[CLI]: Creating ({}) CLI rst files: {}...".format(
+        len(rst_names), rst_names))
+
+    # Write CLI directive and options for each command-line group
+    for group in rst_names:
+        L = ["{}\n".format(group),
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py:334]: Done!

File: docs/conf.py [hunk 10]
```diff
@@ -225,6 +225,175 @@
...
+
+
+def write_rst_files(comm_name, lib_name, rst_names, doc_folder):
+    """Write a rst file with CLI directives for each rst file name.
+
+    Args:
+        comm_name: Command name used in the documentation.
+        lib_name: CLI command library name.
+        rst_names: The rst file names, also used for the command name.
+        doc_folder: The path where cli docs will be saved.
```
[the remark this answers is not in the document]
Comment (santiagogaray) [reply docs/conf.py]: Thanks for this comment, very helpful. There was a bit confusing naming on some of these parameters, mostly given some ambiguity on these variables as they can mean different things in the process. I adjusted the overall variable naming throughout this routine, hopefully this is more clear. I did also add some examples, hope this is what you were asking.

**Comment by santiagogaray** [comment]
Thanks @chriswmackey for your thoughtful comments. I did the changes as suggested and while I was at it I worked a bit more on the code trying to make it a bit more robust. Please let me know if you have any comments when you have a chance,.

I will reply more tomorrow about your post in regards to scope. but just wanted to say that I would be very excited to take on the additional work with the grasshopper plugins.  Thanks!

**Comment by santiagogaray** [comment]
Sorry @chriswmackey, I uploaded, yet again, a newly revised code. It's just some minor improvements that wanted to get in before we merge the file. I forgot also to answer your question about nested groups. I do believe this is not an issue at least in the honeybee-energy case but I will make sure to keep an eye if this is different in other cases. Hope the code looks better but please feel free to make any comments or suggestions if necessary

In regards to the scope of work, thanks for sending all this information about both projects, this is super clear. I will start looking at the repos in the core libraries and estimate the hours to implement the new CLI docs on all of them. The good news though is that with this automated process we should be saving quite some time in this project.
In regards to the grasshopper libraries, as I mentioned yesterday, I will be very interested in working on them as well. Thanks for sending all the relevant information. I did look at the examples posted as well as the latest repo started by Mostapha. I think I understand the overall process but I will probably ask you some specific questions after looking at the grasshopper script. 

Good to have some extra time before we send all of this. I want to make sure I get a good estimate by then, I'll get started with the CLI docs project but feel free to send the script when you get a chance. Thanks again for considering me for this work!
